app/models/katello/concerns/host_managed_extensions.rb

```diff
@@ -66,16 +66,23 @@ def update_os_from_facts
 
         def self.find_with_expiring_pools(_key, _operator, days_from_now)
           minimum_pool_date = (Date.today + days_from_now.to_i.days).strftime('%Y-%m-%d')
+          host_ids = self.joins(:pools).where(["katello_pools.end_date < ?", minimum_pool_date]).distinct.ids
+          if host_ids.blank?
+            return {
+              :conditions => "katello_pools.end_date < ?",
+              :parameter => [minimum_pool_date],
+              :joins => :pools
+            }
+          end
           {
-            :joins => :pools,
-            :conditions => "katello_pools.end_date < ?",
-            :parameter => [minimum_pool_date]
+            :conditions => "hosts.id IN (#{host_ids.join(', ')})"
           }
         end
       end
 
       # show a single host's expiring pools
       def pools_expiring_in_days(days_from_now)
+        return self.pools if days_from_now.blank?
```

Comment:
This handles the situation where the user leaves the "Days from Now" input blank.  It will result in a report with all pools for all hosts.

Reply:
Whenever I see a method like this which refers to an association in the name: *pools*_in_expiring_in_days this tells me "use a scope" that would entail placing a scope definition on the Pool model so we could call `host.pools.expiring_in_days`

That way we have a more generic solution. We could call Katello::Pool.expiring_in_days directly if we wanted!

Reply:
In response to your comment about returning early from this method - I don't think we should do it especially if we're using a scope.

Your template (https://github.com/theforeman/community-templates/pull/705/files#diff-b757c54cdf8d66853a76e863d2269309R23) could be updated to look like this

```
pools = host.pools
pools = pools.expiring_in_days(days_from_now) if days_from_now

pools.each do |pool|
...
```